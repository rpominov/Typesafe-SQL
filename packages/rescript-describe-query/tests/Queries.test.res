open Jest

let then = (promise, fn) => Js.Promise.then_(fn, promise)

let client = ref(None)

beforeAllAsync(() =>
  Client.make()->then(result => {
    switch result {
    | Ok(c) => {
        client := Some(c)
        Promise.resolve()
      }
    | Error(e) => {
        Errors.Loggable.toString(e)->Js.Console.error
        Promise.reject(Promise.makeJsError("Could not create a client"))
      }
    }
  })
)

afterAllAsync(() => {
  client.contents->getExn(__LOC__)->Client.terminate
})

eachAsync(
  [
    "SELECT 1",
    "SELECT 1;",
    "SELECT 1 -- comment",
    "SELECT oid FROM pg_type",
    "SELECT oid FROM pg_type WHERE pg_type = $1",
  ],
  "Describe: %s",
  query => {
    expectAssertions(1)
    client.contents
    ->getExn(__LOC__)
    ->Client.describe(query)
    ->then(description => {
      expect(description)->toMatchSnapshot
      Js.Promise.resolve()
    })
  },
)

testAsync("tableColumn", () => {
  expectAssertions(1)
  client.contents
  ->getExn(__LOC__)
  ->Client.describe("SELECT oid FROM pg_type")
  ->then(description => {
    ((description->getOkExn(__LOC__)).row->getExn(__LOC__)->arrGetExn(0, __LOC__)).tableColumn
    ->getExn(__LOC__)
    ->expect
    ->toEqual({
      attcollation: Some(0),
      attfdwoptions: None,
      attname: Some("oid"),
      attndims: Some(0),
      attnotnull: Some(true),
      attnum: Some(1),
      attoptions: None,
      attrelid: Some(1247),
      atttypid: Some(26),
      atttypmod: Some(-1),
      relname: Some("pg_type"),
    })
    Js.Promise.resolve()
  })
})

testAsync("queue", () => {
  expectAssertions(3)

  let client = client.contents->getExn(__LOC__)

  let a = client->Client.describe("SELECT 1 a")
  let b = client->Client.describe("SELECT 1 b")
  let c = client->Client.describe("SELECT 1 c")

  Js.Promise.all3((a, b, c))->then(((a, b, c)) => {
    expect(((a->getOkExn(__LOC__)).row->getExn(__LOC__)->arrGetExn(0, __LOC__)).name)->toEqual("a")
    expect(((b->getOkExn(__LOC__)).row->getExn(__LOC__)->arrGetExn(0, __LOC__)).name)->toEqual("b")
    expect(((c->getOkExn(__LOC__)).row->getExn(__LOC__)->arrGetExn(0, __LOC__)).name)->toEqual("c")
    Js.Promise.resolve()
  })
})

// TODO: write a query that would contain all kinds of types: Array, Enum etc.
testAsync("complicated types", () => {
  expectAssertions(1)
  client.contents
  ->getExn(__LOC__)
  ->Client.describe("SELECT typacl FROM pg_type")
  ->then(description => {
    (description->getOkExn(__LOC__)).row
    ->getExn(__LOC__)
    ->Js.Array2.map(x => x.dataType)
    ->expect
    ->toMatchSnapshot
    Js.Promise.resolve()
  })
})
