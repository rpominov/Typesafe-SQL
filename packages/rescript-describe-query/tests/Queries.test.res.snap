// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Describe: SELECT 1 -- comment 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "parameters": Array [],
    "row": Array [
      Object {
        "dataType": Object {
          "byVal": true,
          "category": "N",
          "isDefined": true,
          "isPreferred": false,
          "len": 4,
          "name": "int4",
          "namespace": "pg_catalog",
          "oid": 23,
          "typeSpecificData": 0,
          "typeType": "b",
        },
        "name": "?column?",
        "tableColumn": undefined,
      },
    ],
  },
}
`;

exports[`Describe: SELECT 1 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "parameters": Array [],
    "row": Array [
      Object {
        "dataType": Object {
          "byVal": true,
          "category": "N",
          "isDefined": true,
          "isPreferred": false,
          "len": 4,
          "name": "int4",
          "namespace": "pg_catalog",
          "oid": 23,
          "typeSpecificData": 0,
          "typeType": "b",
        },
        "name": "?column?",
        "tableColumn": undefined,
      },
    ],
  },
}
`;

exports[`Describe: SELECT 1; 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "parameters": Array [],
    "row": Array [
      Object {
        "dataType": Object {
          "byVal": true,
          "category": "N",
          "isDefined": true,
          "isPreferred": false,
          "len": 4,
          "name": "int4",
          "namespace": "pg_catalog",
          "oid": 23,
          "typeSpecificData": 0,
          "typeType": "b",
        },
        "name": "?column?",
        "tableColumn": undefined,
      },
    ],
  },
}
`;

exports[`Describe: SELECT oid FROM pg_type 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "parameters": Array [],
    "row": Array [
      Object {
        "dataType": Object {
          "byVal": true,
          "category": "N",
          "isDefined": true,
          "isPreferred": true,
          "len": 4,
          "name": "oid",
          "namespace": "pg_catalog",
          "oid": 26,
          "typeSpecificData": 0,
          "typeType": "b",
        },
        "name": "oid",
        "tableColumn": Object {
          "attcollation": 0,
          "attfdwoptions": undefined,
          "attname": "oid",
          "attndims": 0,
          "attnotnull": true,
          "attnum": 1,
          "attoptions": undefined,
          "attrelid": 1247,
          "atttypid": 26,
          "atttypmod": -1,
          "relname": "pg_type",
        },
      },
    ],
  },
}
`;

exports[`Describe: SELECT oid FROM pg_type WHERE pg_type = $1 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "parameters": Array [
      Object {
        "byVal": false,
        "category": "P",
        "isDefined": true,
        "isPreferred": false,
        "len": -1,
        "name": "record",
        "namespace": "pg_catalog",
        "oid": 2249,
        "typeSpecificData": 1,
        "typeType": "p",
      },
    ],
    "row": Array [
      Object {
        "dataType": Object {
          "byVal": true,
          "category": "N",
          "isDefined": true,
          "isPreferred": true,
          "len": 4,
          "name": "oid",
          "namespace": "pg_catalog",
          "oid": 26,
          "typeSpecificData": 0,
          "typeType": "b",
        },
        "name": "oid",
        "tableColumn": Object {
          "attcollation": 0,
          "attfdwoptions": undefined,
          "attname": "oid",
          "attndims": 0,
          "attnotnull": true,
          "attnum": 1,
          "attoptions": undefined,
          "attrelid": 1247,
          "atttypid": 26,
          "atttypmod": -1,
          "relname": "pg_type",
        },
      },
    ],
  },
}
`;

exports[`complicated types 1`] = `
Array [
  Object {
    "byVal": false,
    "category": "A",
    "isDefined": true,
    "isPreferred": false,
    "len": -1,
    "name": "_aclitem",
    "namespace": "pg_catalog",
    "oid": 1034,
    "typeSpecificData": Object {
      "TAG": 0,
      "delim": ",",
      "elemType": Object {
        "byVal": false,
        "category": "U",
        "isDefined": true,
        "isPreferred": false,
        "len": 12,
        "name": "aclitem",
        "namespace": "pg_catalog",
        "oid": 1033,
        "typeSpecificData": 0,
        "typeType": "b",
      },
    },
    "typeType": "b",
  },
]
`;
